core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -447,7 +450,16 @@ private[spark] class Executor(
       }
     }
 
-    val message = Heartbeat(executorId, tasksMetrics.toArray, env.blockManager.blockManagerId)
+    env.blockTransferService.getMemMetrics(this.executorMetrics)
+    val executorMetrics = if (isLocal) {
+      Utils.deserialize[ExecutorMetrics](Utils.serialize(this.executorMetrics))
```

Comment:
This is due to [SPARK-3465](https://issues.apache.org/jira/browse/SPARK-3465). Currently we do not have any aggregation operations for `ExecutorMetrics`, we can remove this. We can add it back when we do some aggregation.